tensorboard/components/tf_paginated_view/tf-paginated-view2.html

```diff
@@ -0,0 +1,414 @@
+        computed: '_computePageCount(items, _limit)',
+      },
+      _multiplePagesExist: {
+        type: Boolean,
+        computed: '_computeMultiplePagesExist(_pageCount)',
+      },
+      _hasPreviousPage: {
+        type: Boolean,
+        computed: '_computeHasPreviousPage(_currentPage)',
+      },
+      _hasNextPage: {
+        type: Boolean,
+        computed: '_computeHasNextPage(_currentPage, _pageCount)',
+      },
+      _inputWidth: {
+        type: String,
+        computed: '_computeInputWidth(_pageCount)',
+      },
+
+      _pageInputValue: {
+        type: String,  // value displayed in the input field at any time
+        computed: '_computePageInputValue(_pageInputFocused, _pageInputRawValue, _currentPage)',
+        observer: '_updatePageInputValue',
+      },
+      _pageInputRawValue: {
+        type: String,  // updated live as the user types
+        value: '',
+      },
+      _pageInputFocused: {
+        type: Boolean,
+        value: false,
+      },
+      _bootstrapped: {
+        type: Boolean,
+        value: false,
+      },
+    },
+    observers: [
+      '_clampActiveIndex(items)',
+      '_updateRenderedItems(items, _limit, _activeIndex, _pageCount)',
+      '_updateDom(_renderedItems.*)',
+    ],
+
+    behaviors: [
+      tf_dashboard_common.ArrayUpdateHelper,
+      Polymer.Templatizer,
+    ],
+
+    ready() {
+      this.templatize(this.$.template.querySelector('template'));
+    },
+
+    attached() {
+      this._limitListener = () => {
+        this.set('_limit', tf_paginated_view.getLimit());
+      };
+      tf_paginated_view.addLimitListener(this._limitListener);
+      this._limitListener();
+      this._bootstrapDom();
+      this._bootstrapped = true;
+    },
+
+    detached() {
+      tf_paginated_view.removeLimitListener(this._limitListener);
+      this._bootstrapped = false;
+    },
+
+    _updateRenderedItems(items, limit, activeIndex, pageCount) {
+      const activePageIndex = Math.floor(activeIndex / limit);
+      const domItems = (items || [])
+          .slice(activePageIndex * limit, (activePageIndex + 1) * limit);
+      this.updateArrayProp('_renderedItems', domItems, this.getItemKey);
+    },
+
+    _bootstrapDom() {
+      Array.from(this.$.items.children)
+          .forEach(child => Polymer.dom(this.$.items).removeChild(child));
+      this._renderedItems
+          .forEach((item, index) => this._insertItem(item, index));
+    },
+
+    _updateDom(event) {
+      if (!this._bootstrapped) return;
+      if (event.path !== '_renderedItems.splices') return;
+      event.value.indexSplices.forEach(splice => {
+        const {index, addedCount, object, removed} = splice;
+        removed.forEach(() => {
+          Polymer.dom(this.$.items).removeChild(this.$.items.children[index]);
+        });
+        object.slice(index, index + addedCount)
+            .forEach((item, ind) => this._insertItem(item, index + ind));
+      });
+    },
+
+    _insertItem(item, atIndex) {
```

Comment:
I thought you'd mentioned something about dom-repeat with a predicate?  It looks like this approach manually stamps the template as needed.  Which is reasonable if that's what we have to do, but it means that the ArrayUpdateHelper is being used sort of indirectly because we then have to parse the updated slices back out of the complex observer change record, and use them to manually do the right manipulation of the children.  If we were able to still use dom-repeat then it'd seem like we were taking more advantage of ArrayUpdateHelper.

Reply:
I couldn't find a good way to propagate slotted template to a dom-repeat :(
Maybe iron-list had a separate requirement but this pattern of stamping your own DOM seems somewhat natural: https://github.com/PolymerElements/iron-list/blob/1.x/iron-list.html